src/hammer_config_test/test.py

```diff
@@ -935,5 +935,156 @@ def test_meta_deepsubst_subst(self) -> None:
+    def test_wrong_types(self) -> None:
+        """
+        Test that types are checked when retrieving settings.
+        """
+        db = hammer_config.HammerDatabase()
+        base = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: [1, 2, 3]
+        dac: 0
+        wrong: 0
+""", is_yaml=True)
+        base_types = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: list[int]
+        dac: int
+        wrong: list[dict[str, str]]
+""", is_yaml=True)
+
+        db.update_core([base])
+        db.update_types(base_types)
+
+        self.assertEqual(db.get_setting("foo.bar.adc"), [1, 2, 3])
+        with self.assertRaises(TypeError):
+            db.get_setting("foo.bar.wrong")
+
+    def test_wrong_constraints(self) -> None:
+        """
+        Test that custom constraints are checked when retrieving settings.
+        """
+        db = hammer_config.HammerDatabase()
+        base = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: [{name: \"hi\", pin: \"vdd\"}]
+        wrong: [{name: \"hi\", pin: \"vdd\"}]
+""", is_yaml=True)
+        base_types = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: list[dict[str, str]]
+        wrong: int
+""", is_yaml=True)
+
+        db.update_core([base])
+        db.update_types(base_types)
+
+        self.assertEqual(db.get_setting("foo.bar.adc"), [{"name": "hi", "pin": "vdd"}])
+        with self.assertRaises(TypeError):
+            db.get_setting("foo.bar.wrong")
+
+    def test_optional_constraints(self) -> None:
+        """
+        Test that optional constraints are ignored properly.
+        """
+        db = hammer_config.HammerDatabase()
+        base = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: 
+        quu: hi
+""", is_yaml=True)
+        base_types = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: Optional[str]
+        quu: Optional[str]
+""", is_yaml=True)
+
+        db.update_core([base])
+        db.update_types(base_types)
+        
+        self.assertIsNone(db.get_setting("foo.bar.adc"))
+        self.assertEqual(db.get_setting("foo.bar.quu"), "hi")
+
+    def test_decimal_constraints(self) -> None:
+        """
+        Test that Decimal constraints are checked properly.
+        """
+        db = hammer_config.HammerDatabase()
+        base = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: 3.14
+        quu: 69
+""", is_yaml=True)
+        base_types = hammer_config.load_config_from_string("""
+foo:
+    bar:
+        adc: float
+        quu: float
+""", is_yaml=True)
+
+        db.update_core([base])
+        db.update_types(base_types)
+        
+        self.assertEqual(db.get_setting("foo.bar.adc"), 3.14)
+        with self.assertRaises(TypeError):
+            db.get_setting("foo.bar.quu")
+
+    def test_default_types(self) -> None:
+        """
+        Test that all HAMMER defaults are properly type-checked.
```

Comment:
This is great to have for all default keys in Hammer. Can this be made into a method inheritable by all `HammerTool`s and `HammerTechnology`s so that we can auto check in plugin repos? If so, this should also just generate a warning if any `defaults_types.yml` is not found in conjunction with a `defaults.yml` file telling the user to implement it.

Reply:
Could you elaborate on this a little more? Should we check types as soon as they are updated into a given `HammerDatabase`?

Reply:
Oh, I see, currently this is only run when you do unit tests, right? If so, yes, I think the types should be checked whenever a `HammerDatabase` is created/updated.